File: java/383_Ransom_Note.java
/*
https://leetcode.com/problems/ransom-note/
Given an arbitrary ransom note string and another string containing letters from all the magazines, write a function
that will return true if the ransom note can be constructed from the magazines ; otherwise, it will return false.
Each letter in the magazine string can only be used once in your ransom note.
You may assume that both strings contain only lowercase letters.

Count magazine letter frequencies in an array.  Decrement count for each letter in ransomNote, returning false if
any letter count is negative.
Alternatively if unknown set of characters use hashmap as counter.
Time - O(m), magazine length
Space - O(1)
*/

public class Solution {
    public boolean canConstruct(String ransomNote, String magazine) {

        if (ransomNote.length() > magazine.length())
            return false;

        int[] counter = new int[26];

        for (int i = 0; i < magazine.length(); ++i)
            ++counter[(int) magazine.charAt(i) - (int) 'a'];

        for (int j = 0; j < ransomNote.length(); ++j) {
            if (--counter[(int) ransomNote.charAt(j) - (int) 'a'] == -1)
                return false;
        }

        return true;
    }
}

File: python_1_to_1000/383_Ransom_Note.py
_author_ = 'jake'
_project_ = 'leetcode'

# https://leetcode.com/problems/ransom-note/
# Given an arbitrary ransom note string and another string containing letters from all the magazines, write a function
# that will return true if the ransom note can be constructed from the magazines ; otherwise, it will return false.
# Each letter in the magazine string can only be used once in your ransom note.
# You may assume that both strings contain only lowercase letters.

# Count all chars in ransom note asd magazines. If any char is in ransom note but not magazine or has a lower count in
# magazine then return False.
# Time - O(m + n)
# Space - O(1)

from collections import Counter

class Solution(object):
    def canConstruct(self, ransomNote, magazine):
        """
        :type ransomNote: str
        :type magazine: str
        :rtype: bool
        """
        mag_count = Counter(magazine)
        ransom_count = Counter(ransomNote)

        for c in ransom_count:
            if c not in mag_count or mag_count[c] < ransom_count[c]:
                return False

        return True